src/core/LinkedModelGroups.cpp

```diff
@@ -0,0 +1,362 @@
+		}
+	}
+}
+
+
+
+
+void LinkedModelGroup::saveLinksEnabled(QDomDocument &doc, QDomElement &that)
+{
+	if (m_linkEnabled.size())
+	{
+		std::size_t count = 0;
+		for (BoolModel* bmo : m_linkEnabled)
+		{
+			//that.setAttribute(m_models[count++].m_name, bmo->value());
+			bmo->saveSettings(doc, that, m_models[count++].m_name);
+		}
+	}
+}
+
+
+
+
+void LinkedModelGroup::loadValues(const QDomElement &that,
+	const LinkedModelGroup* lmg0)
+{
+	for (ModelInfo& minf : m_models)
+	{
+		std::size_t idx = 0;
+		if (this == lmg0) { idx = lmg0->models().size(); } // force loading
+		else
+		{
+			for (; idx < lmg0->models().size(); ++idx)
+			{
+				if (lmg0->models()[idx].m_name == minf.m_name)
+				{
+					break;
+				}
+			}
+		}
+		if (idx < lmg0->models().size() &&
+			lmg0->linkEnabledModel(idx)->value())
+		{
+			// link is enabled => it will load automatically
+		}
+		else
+		{
+			// try to load, if it fails, this will load a sane initial value
+			minf.m_model->loadSettings(that, minf.m_name);
+		}
+	}
+}
+
+
+
+
+void LinkedModelGroup::loadLinksEnabled(const QDomElement &that)
+{
+	if (m_linkEnabled.size())
+	{
+		std::size_t count = 0;
+		for (BoolModel* bmo : m_linkEnabled)
+		{
+			bmo->loadSettings(that, m_models[count++].m_name);
+		}
+	}
+}
+
+
+
+
+void LinkedModelGroup::linkStateChangedSlot()
+{
+	QObject* sender = QObject::sender();
+	Q_ASSERT(sender);
+	BoolModel* bmo = qobject_cast<BoolModel*>(sender);
+	Q_ASSERT(bmo);
+	int modelNo = -1, count = 0;
+	for (BoolModel* bmo2 : m_linkEnabled)
+	{
+		if (bmo2 == bmo) { modelNo = count; }
+		++count;
+	}
+	Q_ASSERT(modelNo >= 0);
+	emit linkStateChanged(modelNo, bmo->value());
+}
+
+
+
+
+void LinkedModelGroup::addModel(AutomatableModel *model, const QString &name)
+{
+	m_models.emplace_back(name, model);
+}
+
+
+
+
+/*
+	LinkedModelGroups
+*/
+
+
+
+
+LinkedModelGroups::~LinkedModelGroups() {}
+
+
+
+
+void LinkedModelGroups::createMultiChannelLinkModel()
+{
+	m_multiChannelLinkModel.reset(new BoolModel(true, nullptr));
+}
+
+
+
+
+void LinkedModelGroups::linkPort(int port, bool state)
+{
+	LinkedModelGroup* first = getGroup(0);
+	LinkedModelGroup* cur;
+
+	if (state)
+	{
+		for (std::size_t i = 1; (cur=getGroup(i)); ++i)
+		{
+			first->linkControls(cur, port);
+		}
+	}
+	else
+	{
+		for (std::size_t i = 1; (cur=getGroup(i)); ++i)
+		{
+			first->unlinkControls(cur, port);
+		}
+
+		// m_multiChannelLinkModel.setValue() will call
+		// updateLinkStatesFromGlobal()...
+		// m_noLink will make sure that this will not unlink any other ports
+		m_noLink = true;
+		m_multiChannelLinkModel->setValue( false );
+	}
+}
+
+
+
+
+void LinkedModelGroups::updateLinkStatesFromGlobal()
+{
+	LinkedModelGroup* first = getGroup(0);
+	if (m_multiChannelLinkModel->value())
+	{
+		first->linkAllModels(true);
+	}
+	else if (!m_noLink)
+	{
+		first->linkAllModels(false);
+	}
+
+	// if global channel link state has changed, always ignore link
+	// status of individual ports in the future
+	m_noLink = false;
```

Comment:
The comment says "if ... ", but doesn't this happen every time the function is called? Maybe I'm misunderstanding.

Reply:
That was copied from LADSPA :smile: Right now, I think the other comments about `m_noLink` make clear what the variable is used for. I'd remove the whole comment, if you don't veto.

Reply:
No complaints here!